README.md
# My Dev-C++

![](https://raw.githubusercontent.com/skylee03/My-Dev-Cpp/master/devcpp.ico)

*A new and improved fork of [Orwell Dev-C++](https://sourceforge.net/projects/orwelldevcpp/)*

[![Version: v5.13](https://img.shields.io/badge/Version-v5.13-green)](https://github.com/skylee03/My-Dev-Cpp/releases/)
[![License: GPL v2](https://img.shields.io/badge/License-GPL%20v2-green.svg)](https://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html)
![Platform: Windows](https://img.shields.io/badge/Platform-Windows-green)
![Made with Delphi 7](https://img.shields.io/badge/Made_with-Delphi_7-green)

## Main Features

- [MinGW-W64](https://sourceforge.net/projects/mingw-w64/) 8.1.0 32/64-bit Compiler
- Syntax highlighting
- Code completion
- Code insight
- Editable shortcuts
- GPROF profiling
- GDB debugging
- [AStyle](http://astyle.sourceforge.net/) code formatting
- Devpak IDE extensions
- External tools

## Screenshots
![Compiling](https://i.loli.net/2019/12/07/IlD5Z68xVvmdcSn.png)
![Debugging](https://i.loli.net/2019/12/07/xAK2PJrYNQ7a5ig.png)

## Old Developers
[![bloodshed.gif](https://i.loli.net/2019/12/21/zH5n8UOfk2YVEtb.gif)](http://www.bloodshed.net/)

NEWS.txt
Version 5.14 - 5 July 2020

- Restore the status of TDM-GCC as the default compiler version, since its maintaince has been restart.
- Upgrade the default compiler to TDM-GCC 9.2.0.

Version 5.13 - 20 January 2020

- Fixed a bug in console pauser related to spaces in paths and parameters.
- Add newer C/C++ standards flags.
- Recover splash form which has been removed since v5.10.
- Fixed some translation files (zh_cn/zh_tw/en).

Version 5.12 - 7 December 2019

- Upgraded the default compiler to MinGW-W64 GCC 8.1.0.
- Improve the naming mechanism for auto-configured compiler sets by automatically assembling user-friendly names for MinGW-W64 compilers.
- Fixed some miskates in the English translation.
- Add Chinese (Simplified/Traditional) translation of the setup wizard.
- Add new contents into the "About".

=========================

Version 5.11 - 27 April 2015

- Fixed crash related to double clicking on a compiler error when a selection was made.
- Upgraded the default compiler to TDM-GCC 4.9.2.
- Improved startup speed.
- Fixed Abort Compilation button not working anymore.
- Fixed crash in TCppParser.CheckForTypedefStruct.
- Fixed crash in TCppParser.HandleEnum.
- Fixed some typos in the English translation (thanks to Hiro5).
- Updated the Catalan translation (thanks to Hiro5).
- Updated the Czech translation (thanks to tringi).
- Fixed some hickups in the build process of Dev-C++ itself.

Version 5.10 - 11 March 2015

- Improved startup speed.
- Removed splash window.
- Temporarily removed the web update window.
- Rewrote the LangCheck tool that can be used to validate translation files.
- Improved indent guides painting and positioning.
- Removed full screen information bar.
- Added GUI option for maximum line length option to AStyle > Formatter Options.
- Fixed parameter mismatch in Chinese translation.
- Fixed memory leak in Edit > Unindent.
- Improved behaviour of Edit > Toggle Comment.
- Improved opening speed of Tools > Compiler Options.

Version 5.9.2 - 1 Februari 2015

- Changed Format Current File shortcut from Shift+F to Shift+Ctrl+A.
- Removed/Changed all default Alt+(Key) shortcuts because they interfere with Alt menu navigation.
- Assigned some new default shortcuts like Ctrl+B (Open Containing Folder) and F2 (Rename File).
- Fixed Show Makefile being executable when no compiler set is configured.

Version 5.9.1 - 31 Januari 2015

- Updated Tools >> Edit Shortcuts.
- Removed Ctrl+Alt combination in default shortcuts to ensure that the AltGr button keeps functioning:
http://blogs.msdn.com/b/oldnewthing/archive/2004/03/29/101121.aspx

- Fixed formatter options command line not showing up after first launch.
- Added source LaTeX of testing document DevTest.pdf.
- Widened the Menu >> Search >> Find window in order to accomodate longer translations.
- Fixed a crash when launching for the first time with no compilers available.
- Added last selection arrow to Find Results.
- Fixed out of bounds error in TCppParser.HandleMethod.

Version 5.9.0 - 5 Januari 2015

- Added AStyle integration.
- Remodeled Help >> About window.
- When looking up a piece of code by Ctrl+Clicking the destination line is centered.
- Removed caching from Tools >> Editor Options.
- Widened the exception window.
- Removed caching from the first-time configuration window.
- Improved parsing progress messages in the status bar.
- Improved performance of the classes list toolbar.
- Improved function declaration/definition switching.
- Significantly improved performance when selecting text.
- Fixed parser error when parsing C-style cast of string constant.
- Huge rewrite of code folding implementation.
- Major improvements in typing responsiveness.
- Major reduction in flickering during typing.
- Around 40-50% reduction in memory usage per open file.
- Fixed devcppPortable not passing files with spaces in names to devcpp correctly.
- Improved performance when opening multiple files from explorer using Dev-C++.
- Code completion does not show up anymore for periods typed as part of numbers.

Version 5.8.3 - 10 November 2014

- Slightly decreased flicker during editor opening.
- Makefiles are now edited as if they are source files. It's better than nothing.
- Opening braces after default are now completed correctly.
- Fixed a bug in NewFunctionFrm and NewVarFrm that caused crashes (thank you for reporting).
- Rewritten ancient source code of devcppPortable.exe. It is now immune to overflows due to arguments of length more than 400.
- Fixed a bug in TCppParser that caused it to ignore project include paths.
- Fixed a crash in TCppTokenizer due to spaces before #include in combination with comments after the <file> or "file" part (like " #include <foo> // bar").

Version 5.8.2 - 1 November 2014

- Fixed bug in procedure TMainForm.EditorSaveTimer that corrupted the editor views.
- Updated provided commit command.

Version 5.8.1 - 28 Oktober 2014

- Fixed an overflow error in TCppParser.CheckForStructs that caused errors when opening/saving files.
- Fixed a crash in TEditorList.GetFocusedPageControl that fixes random crashes.
- Refactored symbol completion code.
- The code completion window does not show anymore when the cursor changes during the timer rundown after typing . or :: or ->.
- Improved multiple file opening performance.
- Improved project closing performance.
- Updated compilation readme.

Version 5.8.0 - 21 Oktober 2014

- Added support for two file views next to each other.
- All windows are now opened at the centre of the main window instead of at random places across random monitors.
- Slightly improved performance of the function tooltip.
- Fixed a bug in makefile creation of C DLL's that prevented compilation.
- When compiling, header files are now treated separately from other files.
- Removed CVS support. CVS is considered deprecated, we recommend using explorer based code managers for alternatives like Git or SVN.
- Fixed a crash in the file change monitor code.

Version 5.7.1 - 29 Juli 2014


Source/AboutFrm.pas
{
    This file is part of Dev-C++
    Copyright (c) 2004 Bloodshed Software

    Dev-C++ is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    Dev-C++ is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Dev-C++; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
}

unit AboutFrm;

interface

uses
  Windows, SysUtils, Classes, version, Graphics, Controls, Forms, Dialogs,
  StdCtrls, Buttons, ExtCtrls;

type
  TAboutForm = class(TForm)
    lblVersion: TLabel;
    btnOk: TBitBtn;
    lblCopyright: TLabel;
    grpLicense: TGroupBox;
    txtLicense: TMemo;
    grpLinks: TGroupBox;
    lblBlood: TLabel;
    lblBloodURL: TLabel;
    lblMinGWW64: TLabel;
    lblMinGWW64URL: TLabel;
    lblResources: TLabel;
    lblResourcesURL: TLabel;
    lblBlog: TLabel;
    lblBlogURL: TLabel;
    btnAuthors: TBitBtn;
    btnUpdateCheck: TBitBtn;
    timerFish: TTimer;
    pnlFish: TPanel;
    FishImage: TImage;
    imgBanner: TImage;
    bv2: TBevel;
    lblSubreddit: TLabel;
    lblSubredditURL: TLabel;
    lblRepository: TLabel;
    lblRepositoryURL: TLabel;
    lblPre511: TLabel;
    lblCompilers: TLabel;
    lblPre4992: TLabel;
    bv3: TBevel;
    bv1: TBevel;
    lblPost511: TLabel;
    lblGithub: TLabel;
    Label1: TLabel;
    procedure URLLabelClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure btnAuthorsClick(Sender: TObject);
    procedure btnAuthorsDragOver(Sender, Source: TObject; X, Y: Integer; State: TDragState; var Accept: Boolean);
    procedure btnAuthorsDragDrop(Sender, Source: TObject; X, Y: Integer);
    procedure timerFishTimer(Sender: TObject);
    procedure FishImageClick(Sender: TObject);
    procedure FormClose(Sender: TObject; var Action: TCloseAction);
  private
    procedure LoadText;
  end;

implementation

uses
  ShellAPI, devcfg, utils, MultiLangSupport, main;

{$R *.dfm}

procedure TAboutForm.LoadText;
begin
  // Set interface font
  Font.Name := devData.InterfaceFont;
  Font.Size := devData.InterfaceFontSize;

  Caption := Lang[ID_AB_CAPTION];
  grpLicense.Caption := Lang[ID_AB_LICENSE];
  grpLinks.Caption := Lang[ID_AB_WEBCAP];
  lblPost511.Caption := Lang[ID_AB_POST511];
  lblGithub.Caption := Lang[ID_AB_GITHUB];
  lblMinGWW64.Caption := Lang[ID_AB_TDMGCC];
  lblPre511.Caption := Lang[ID_AB_PRE511];
  lblRepository.Caption := Lang[ID_AB_REPOSITORY];
  lblSubreddit.Caption := Lang[ID_AB_SUBREDDIT];
  lblBlog.Caption := Lang[ID_AB_BLOGNAME];
  lblCompilers.Caption := Lang[ID_AB_COMPILERS];
  lblPre4992.Caption := Lang[ID_AB_PRE4992];
  lblBlood.Caption := Lang[ID_AB_LBLBLOODSITE];
  lblResources.Caption := Lang[ID_AB_LBLEMAIL];
  btnOk.Caption := Lang[ID_BTN_OK];
  btnUpdateCheck.Caption := Lang[ID_AB_UPDATE];
  btnAuthors.Caption := Lang[ID_BTN_AUTHOR];
end;

procedure TAboutForm.URLLabelClick(Sender: TObject);
var
  S: AnsiString;
begin
  S := TLabel(Sender).Caption;
  ShellExecute(GetDesktopWindow(), 'open', PAnsiChar(s), nil, nil, SW_SHOWNORMAL);
end;

procedure TAboutForm.FormCreate(Sender: TObject);
var
  datestring: AnsiString;
begin
  LoadText;

  // Use modified time, not the PE headers
  datestring := GetBuildTime(ParamStr(0));
  lblVersion.Caption := lblVersion.Caption + DEVCPP_VERSION + #13#10 + 'Build time: ' + datestring;

  if FileExists(devData.Splash) then // TODO: check all folders?
    imgBanner.Picture.LoadFromFile(devData.Splash);
end;

procedure TAboutForm.btnAuthorsClick(Sender: TObject);
const
  MessageText =
    'Authors:'#13#10#13#10 +
    '- Post-5.1.1 development: Mingtian Yang'#13#10 +
    '- Pre-5.1.1 development: Johan Mes'#13#10 +
    '- Pre-4.9.9.2 development: Colin Laplace, Mike Berg, Hongli Lai, Yiannis Mandravellos'#13#10 +
    '- Contributors: Peter Schraut, Marek Januszewski, Anonymous'#13#10 +
    '- MinGW compiler system: Mumit Khan, J.J. van der Heijden, Colin Hendrix and GNU developers'#13#10 +
    '- Splash screen and association icons: Matthijs Crielaard'#13#10 +
    '- New Look theme: Gerard Caulfield'#13#10 +
    '- Gnome icons: Gnome designers'#13#10 +
    '- Blue theme: Thomas Thron';
begin
  MessageBeep($F);
  MessageDlg(MessageText, MtInformation, [MbOK], 0);
end;

procedure TAboutForm.btnAuthorsDragOver(Sender, Source: TObject; X,
  Y: Integer; State: TDragState; var Accept: Boolean);
begin
  Accept := True;
end;

procedure TAboutForm.btnAuthorsDragDrop(Sender, Source: TObject; X,
  Y: Integer);
begin
  pnlFish.Left := 0 - pnlFish.Width;
  pnlFish.Top := Random(Height - pnlFish.Height - (pnlFish.Height div 3));
  pnlFish.Tag := 0;
  timerFish.Enabled := True;
end;

procedure TAboutForm.timerFishTimer(Sender: TObject);
begin
  if ((pnlFish.Tag = 0) and (pnlFish.Left > Width + 10)) or
    ((pnlFish.Tag <> 1) and (pnlFish.Left < 0 - pnlFish.Width - 10)) then
    timerFish.Enabled := False;

  if pnlFish.Tag = 0 then
    pnlFish.Left := pnlFish.Left + 5
  else
    pnlFish.Left := pnlFish.Left - 5;
  pnlFish.Top := pnlFish.Top + 1;
end;

procedure TAboutForm.FishImageClick(Sender: TObject);
begin
  pnlFish.Tag := not pnlFish.Tag;
end;

procedure TAboutForm.FormClose(Sender: TObject; var Action: TCloseAction);
begin
  Action := caFree;
end;

end.



Source/SplashFrm.pas
{
    This file is part of Dev-C++
    Copyright (c) 2004 Bloodshed Software

    Dev-C++ is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    Dev-C++ is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Dev-C++; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
}

unit SplashFrm;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  ExtCtrls, ComCtrls, Version;

type
  TSplashForm = class(TForm)
    Image: TImage;
    Statusbar: TStatusbar;
    procedure FormCreate(Sender: TObject);
    procedure FormClose(Sender: TObject; var Action: TCloseAction);
  public
    procedure SetText(const LoadingText: AnsiString);
  end;

var
  SplashForm: TSplashForm = nil;

implementation

uses
  devcfg;

{$R *.dfm}

procedure TSplashForm.FormCreate(Sender: TObject);
begin
  if (devData.Splash <> '') and FileExists(devData.Splash) then begin
    Image.Picture.LoadFromFile(devData.Splash);
    ClientWidth := Image.Width;
    ClientHeight := Image.Height + Statusbar.Height;
  end;
  Show;
  Update;
end;

procedure TSplashForm.FormClose(Sender: TObject; var Action: TCloseAction);
begin
  Action := caFree;
end;

procedure TSplashForm.SetText(const LoadingText: AnsiString);
begin
  Statusbar.SimpleText := 'Dev-C++ ' + DEVCPP_VERSION + ' ' + LoadingText;
end;

end.



Source/MultiLangSupport.pas
{
    This file is part of Dev-C++
    Copyright (c) 2004 Bloodshed Software

    Dev-C++ is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    Dev-C++ is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Dev-C++; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
}

unit MultiLangSupport;

interface

uses
  Windows, Dialogs, SysUtils, Classes;

{$I LangIDs.inc}

type
  TdevMultiLangSupport = class(TObject)
  private
    fLangList: TStringList;
    fLangListLoaded: Boolean;
    fLangFile: AnsiString;
    fCurLang: AnsiString;
    fStrings: TStringList;
    fDefaultLangStrings: TStringList;
    function GetString(ID: integer): AnsiString;
    function GetLangName: AnsiString;
    constructor Create;
  public
    destructor Destroy; override;
    procedure SelectLanguage;
    procedure CheckLanguageFiles;
    procedure Open(const FileName: AnsiString);
    procedure SetLang(const Lang: AnsiString);
    function GetLangList: TStringList;
    function FileFromDescription(const Desc: AnsiString): AnsiString;
    property Strings[index: integer]: AnsiString read GetString; default;
    property CurrentLanguage: AnsiString read GetLangName;
    property Langs: TStringList read GetLangList;
  end;

function Lang: TdevMultiLangSupport;

implementation

uses
  LangFrm, Forms, Utils, Version, Controls, devCFG;

var
  fLangSingleton: TdevMultiLangSupport = nil;

function Lang: TdevMultiLangSupport;
begin
  if not Assigned(fLangSingleton) and not Application.Terminated then
    fLangSingleton := TdevMultiLangSupport.Create;
  Result := fLangSingleton;
end;

constructor TdevMultiLangSupport.Create;
var
  DefaultLangFile: AnsiString;
begin
  inherited;
  fLangList := TStringList.Create;
  fLangListLoaded := False; // only load when needed
  fStrings := TStringList.Create;
  fDefaultLangStrings := TStringList.Create;

  // Load default english file from languages directory
  DefaultLangFile := ValidateFile('en.lng', devDirs.Lang);
  if DefaultLangFile = '' then begin
    MessageDlg('Could not open language file en.lng', mtError, [mbOK], 0);
  end else begin
    fDefaultLangStrings.LoadFromFile(DefaultLangFile);
  end;
end;

destructor TdevMultiLangSupport.Destroy;
begin
  FreeAndNil(fLangList);
  FreeAndNil(fStrings);
  FreeAndNil(fDefaultLangStrings);
  fLangSingleton := nil;
  inherited;
end;

procedure TdevMultiLangSupport.Open(const Filename: AnsiString);
var
  LangFile: AnsiString;
begin
  // Load file from languages directory
  LangFile := ValidateFile(Filename, devDirs.Lang);
  if LangFile = '' then begin
    MessageDlg('Could not open language file ' + Filename, mtError, [mbOK], 0);
    Exit;
  end;

  // Load file into fStrings
  fStrings.LoadFromFile(LangFile);
  fLangFile := LangFile;

  // Get languange
  fCurLang := fStrings.Values['Lang'];
  if fCurLang = '' then
    fCurLang := ChangeFileExt(ExtractFileName(LangFile), '');
  devData.Language := ExtractFileName(LangFile);
end;

procedure TdevMultiLangSupport.CheckLanguageFiles;
var
  I: integer;
  s: AnsiString;
  sl: TStringList;
begin
  try
    fLangList.Clear;
    if devDirs.Lang = '' then
      Exit;

    // Load list of all files
    FilesFromWildcard(devDirs.Lang, '*.lng', fLangList, False, False, True);
    if fLangList.Count > 0 then begin
      fLangList.Sort;
      sl := TStringList.Create;
      try
        for I := 0 to fLangList.Count - 1 do begin
          sl.LoadFromFile(fLangList[I]);
          s := sl.Values['Lang'];
          if SameText(ExtractFileName(fLangList[I]), 'en.lng') and (devData.Language = '') then
            fCurLang := s;
          if s = '' then
            fLangList[I] := format('%s=%s', [fLangList[I], ChangeFileExt(ExtractFileName(fLangList[I]), '')])
          else
            fLangList[I] := format('%s=%s', [fLangList[I], s]);
        end;
      finally
        sl.Free;
      end;
    end;
    if fCurLang = '' then
      fCurLang := devData.Language;
  finally
    fLangListLoaded := True;
  end;
end;

function TdevMultiLangSupport.GetString(ID: integer): AnsiString;
begin
  result := fStrings.Values[IntToStr(ID)];
  if Result = '' then
    Result := fDefaultLangStrings.Values[IntToStr(ID)];
  if result = '' then
    result := format('<ID %d translation missing>', [ID])
  else
    result := StringReplace(result, '<CR>', #13#10, [rfReplaceAll]);
end;

function TdevMultiLangSupport.GetLangName: AnsiString;
begin
  result := fCurLang;
end;

procedure TdevMultiLangSupport.SelectLanguage;
begin
  with TLangForm.Create(nil) do try
    ShowModal;
  finally
    Free;
  end;
end;

procedure TdevMultiLangSupport.SetLang(const Lang: AnsiString);
var
  I: integer;
begin
  if SameText(Lang, fCurLang) then
    exit;
  if not fLangListLoaded then
    CheckLanguageFiles; // can be slow, so load on demand
  for I := 0 to fLangList.Count - 1 do
    if SameText(ExtractFileName(fLangList.Names[I]), Lang) then begin
      Open(fLangList.Names[I]);
      break;
    end;
end;

function TdevMultiLangSupport.FileFromDescription(const Desc: AnsiString): AnsiString;
var
  i: integer;
begin
  // returns the filename of the lang file described as Desc
  // for example with Desc="English (Original)", returns "en.lng"
  Result := Desc;
  if not fLangListLoaded then
    CheckLanguageFiles; // can be slow, so load on demand
  for i := 0 to fLangList.Count - 1 do
    if SameText(fLangList.ValueFromIndex[i], Desc) then begin
      Result := ExtractFilename(fLangList.Names[i]);
      Break;
    end;
end;

function TdevMultiLangSupport.GetLangList: TStringList;
begin
  if not fLangListLoaded then
    CheckLanguageFiles; // can be slow, so load on demand
  Result := fLangList;
end;

end.



Source/Tools/ConsolePauser/main.cpp
// Execute & Pause
// Runs a program, then keeps the console window open after it finishes

#include <string>
using std::string;
#include <stdio.h>
#include <windows.h>

#define MAX_COMMAND_LENGTH 32768
#define MAX_ERROR_LENGTH 2048

LONGLONG GetClockTick() {
	LARGE_INTEGER dummy;
	QueryPerformanceCounter(&dummy);
	return dummy.QuadPart;
}

LONGLONG GetClockFrequency() {
	LARGE_INTEGER dummy;
	QueryPerformanceFrequency(&dummy);
	return dummy.QuadPart;
}

void PauseExit(int exitcode) {
	system("pause");
	exit(exitcode);
}

string GetErrorMessage() {
	string result(MAX_ERROR_LENGTH,0);
	
	FormatMessage(
		FORMAT_MESSAGE_FROM_SYSTEM|FORMAT_MESSAGE_IGNORE_INSERTS,
		NULL,GetLastError(),MAKELANGID(LANG_NEUTRAL,SUBLANG_DEFAULT),&result[0],result.size(),NULL);
	
	// Clear newlines at end of string
	for(int i = result.length()-1;i >= 0;i--) {
		if(isspace(result[i])) {
			result[i] = 0;
		} else {
			break;
		}
	}
	return result;
}

string GetCommand(int argc,char** argv) {
	string result;
	for(int i = 1;i < argc;i++) {
		// Quote the argument if it contains spaces
		if(string(argv[i]).find(" ") != string::npos) {
			result += string("\"") + string(argv[i]) + string("\"");
		} else {
			result += string(argv[i]);
		}
		
		// Add a space except for the last argument
		if(i != (argc-1)) {
			result += string(" ");
		}
	}
	
	if(result.length() > MAX_COMMAND_LENGTH) {
		printf("\n--------------------------------");
		printf("\nError: Length of command line string is over %d characters\n",MAX_COMMAND_LENGTH);
		PauseExit(EXIT_FAILURE);
	}
	
	return result;
}

DWORD ExecuteCommand(string& command) {
	STARTUPINFO si;
	PROCESS_INFORMATION pi;
	memset(&si,0,sizeof(si));
	si.cb = sizeof(si);
	memset(&pi,0,sizeof(pi));
	
	if(!CreateProcess(NULL, (LPSTR)command.c_str(), NULL, NULL, false, 0, NULL, NULL, &si, &pi)) {
		printf("\n--------------------------------");
		printf("\nFailed to execute \"%s\":",command.c_str());
		printf("\nError %lu: %s\n",GetLastError(),GetErrorMessage().c_str());
		PauseExit(EXIT_FAILURE);
	}
	WaitForSingleObject(pi.hProcess, INFINITE); // Wait for it to finish
	
	DWORD result = 0;
	GetExitCodeProcess(pi.hProcess, &result);
	return result;
}

int main(int argc, char** argv) {
	
	// First make sure we aren't going to read nonexistent arrays
	if(argc < 2) {
		printf("\n--------------------------------");
		printf("\nUsage: ConsolePauser.exe <filename> <parameters>\n");
		PauseExit(EXIT_SUCCESS);
	}
	
	// Make us look like the paused program
	SetConsoleTitle(argv[1]);
	
	// Then build the to-run application command
	string command = GetCommand(argc,argv);
	
	// Save starting timestamp
	LONGLONG starttime = GetClockTick();
	
	// Then execute said command
	DWORD returnvalue = ExecuteCommand(command);
	
	// Get ending timestamp
	LONGLONG endtime = GetClockTick();
	double seconds = (endtime - starttime) / (double)GetClockFrequency();

	// Done? Print return value of executed program
	printf("\n--------------------------------");
	printf("\nProcess exited after %.4g seconds with return value %lu\n",seconds,returnvalue);
	PauseExit(EXIT_SUCCESS);
}


Source/Version.pas
{
    This file is part of Dev-C++
    Copyright (c) 2004 Bloodshed Software

    Dev-C++ is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    Dev-C++ is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Dev-C++; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
}

unit Version;

interface

//path delimiter
const
{$IFDEF WIN32}
  pd = '\';
{$ENDIF}
{$IFDEF LINUX}
  pd = '/';
{$ENDIF}

  // exe properties
  DEVCPP = 'Dev-C++';
  DEVCPP_VERSION = '5.14';

  // delimiters
  DEV_INTERNAL_OPEN = '$__DEV_INTERNAL_OPEN';

  // config file names
  DEV_DEFAULTCODE_FILE = 'defaultcode.ini';
  DEV_SHORTCUTS_FILE = 'shortcuts.ini';
  DEV_CLASSFOLDERS_FILE = 'classfolders.ini';
  DEV_CODEINS_FILE = 'codeinsertion.ini';
  DEV_WEBMIRRORS_FILE = 'webmirrors.ini';
  DEV_TOOLS_FILE = 'tools.ini';

  // make stuff
  DEV_MAKE_FILE = 'Makefile.win';

  // themes
  DEV_GNOME_THEME = 'Gnome';
  DEV_NEWLOOK_THEME = 'New Look';
  DEV_BLUE_THEME = 'Blue';
  DEV_INTERNAL_THEME = 'New Look';

  LANGUAGE_DIR = 'Lang' + pd;
  ICON_DIR = 'Icons' + pd;
  HELP_DIR = 'Help' + pd;
  TEMPLATE_DIR = 'Templates' + pd;
  THEME_DIR = 'Themes' + pd;
  PACKAGES_DIR = 'Packages' + pd;

  // file extensions
  LIB_EXT = '.a';
  OBJ_EXT = '.o';
  DLL_EXT = '.dll';
  EXE_EXT = '.exe';
  DEV_EXT = '.dev';
  HTML_EXT = '.html';
  RTF_EXT = '.rtf';
  TEX_EXT = '.tex';
  INI_EXT = '.ini';
  TEMPLATE_EXT = '.template';
  SYNTAX_EXT = '.syntax';
  C_EXT = '.c';
  CPP_EXT = '.cpp';
  CC_EXT = '.cc';
  CXX_EXT = '.cxx';
  CP2_EXT = '.c++';
  CP_EXT = '.cp';
  H_EXT = '.h';
  HPP_EXT = '.hpp';
  RC_EXT = '.rc';
  RES_EXT = '.res';
  RH_EXT = '.rh';
  GCH_EXT = '.gch'; // precompiled header
  DEF_EXT = '.def'; // definitions file

  // program defaults
  MAKE_PROGRAM = 'mingw32-make.exe';
  GCC_PROGRAM = 'gcc.exe';
  GPP_PROGRAM = 'g++.exe';
  GDB_PROGRAM = 'gdb.exe';
  GDB32_PROGRAM = 'gdb32.exe';
  WINDRES_PROGRAM = 'windres.exe';
  GPROF_PROGRAM = 'gprof.exe';
  PACKMAN_PROGRAM = 'packman.exe';
  CLEAN_PROGRAM = 'rm.exe'; // cannot use del, it doesn't have a quiet mode

  // File dialog filters
  FLT_ALLFILES = 'All files (*.*)|*.*|';
  FLT_TEXTS = 'Text files (*.txt)|*.txt';
  FLT_LIBRARIES = 'Libraries (*.lib;*.a)|*.lib;*.a';
  FLT_PROJECTS = 'Dev-C++ project (*.dev)|*.dev';
  FLT_HEADS = 'Header files (*.h;*.hpp;*.rh;*.hh)|*.h;*.hpp;*.rh;*.hh';
  FLT_CS = 'C source files (*.c)|*.c';
  FLT_CPPS = 'C++ source files (*.cpp;*.cc;*.cxx;*.c++;*.cp)|*.cpp;*.cc;*.cxx;*.c++;*.cp';
  FLT_RES = 'Resource scripts (*.rc)|*.rc';
  FLT_MSVCPROJECTS = 'MS Visual C++ projects (*.dsp)|*.dsp';
  FLT_CBPROJECTS = 'Code::Blocks projects (*.cbp)|*.cbp';

  // Custom synedit style properties
  cBP = 'Breakpoints';
  cErr = 'Error line';
  cABP = 'Active breakpoints';
  cGut = 'Gutter';
  cSel = 'Selected text';
  cFld = 'Folding lines';

  // GPROF commands and displays
  GPROF_CHECKFILE = 'gmon.out';
  GPROF_CMD_GENFLAT = '-p';
  GPROF_CMD_GENMAP = '-q';

implementation

end.

